<a href="https://colab.research.google.com/github/zxcvnm05/ecommerce_customer_behavior_analysis/blob/main/Data_Preprocessing_group_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Part 1: Topic Selection and Data Collection**

**1. Topic Title:**\
**Analysis of Customer Purchasing Patterns and Return Behavior in E-commerce.**\
**2. Research Question & Managerial Question:**\
**2.1. Managerial Question:**\
How can e-commerce businesses improve customer satisfaction and reduce product return rates by analyzing customer demographics, payment behavior, and delivery performance?\
Justification for the Managerial Question\
Strategic Decision Making: The core objective of e-commerce management is to balance growth with operational efficiency. This question is justified because the dataset is explicitly designed for "Churn & return prediction" and "Sales analytics".\
Impact on Bottom Line: In a real-world business context, high return rates significantly erode profit margins due to reverse logistics costs. By analyzing the intersection of demographics and delivery performance, managers can identify high-risk segments and optimize the supply chain to protect revenue.\
**2.2. Research Questions:**\
**RQ1. How do customer demographics (age and gender) influence total order values?**
* Data Utilization: The dataset includes specific variables for customer_age (18–69), customer_gender, and order_value_usd,.
* Market Segmentation: From a managerial perspective, understanding which demographic groups spend more allows for targeted marketing and personalized recommendations, which is a stated ideal use for this data.\
**RQ2. Does payment method affect purchasing behavior and return likelihood?**
* Behavioral Analysis: With diverse options like UPI, Wallet, and Credit Card, this question explores how financial friction or convenience impacts the finality of a sale.
* Risk Mitigation: Identifying if specific payment methods correlate with the 51% return rate observed in the data allows businesses to adjust checkout incentives or fraud-detection protocols for certain payment types.\
**RQ3. Which product categories experience the highest frequency of returns?**
* Operational Focus: The dataset covers categories such as Beauty, Fashion, Electronics, and Groceries.
* Inventory Quality: This question is essential for "return trends" analysis. If a category like "Fashion" shows a disproportionate return rate, it signals a need for better sizing guides or improved product descriptions to align customer expectations with reality\
**RQ4. What is the relationship between delivery duration and customer ratings?**
* Service Level Analysis: This question directly utilizes the delivery_time_days (ranging from 1 to 14) and the customer_rating (1 to 5),.
* Performance Benchmarking: Since the dataset is designed to analyze "delivery efficiency", this question determines the "threshold of patience" for customers. It helps managers decide if investing in faster shipping (e.g., reducing delivery from 10 days to 5) actually results in a measurable boost in customer satisfaction scores.\
**3. Data Source Link:**
E-Commerce Customer Behavior Dataset. The dataset is hosted on Kaggle and was updated by Muhammad Dastgeer.\
**4. Description of How the Data Was Collected:**\
The data was collected from a public open-source platform [E-Commerce Customer Behavior](https://www.kaggle.com/datasets/dastgeerjutt/e-commerce-customer-behavior/data), Kaggle, in a CSV format. The dataset consists of 5,000 e-commerce order records that capture a variety of metrics including customer demographics, transaction details, and delivery performance.\
**5. Raw Dataset File Information:**
* File Name: ecommerce_customer_behavior_5000.csv
* Dimensions: 5,000 rows and 10 columns.
* Data Variables (Columns):
* order_id: Unique identifier for each order.
* customer_age: Age of the customer (ranging from 18 to 69).
* customer_gender: Gender of the customer.
* product_category: Categories such as Beauty, Fashion, Electronics, etc.
* payment_method: Methods used including Credit Card, Debit Card, Wallet, and UPI.
* order_value_usd: The total value of the order in USD.
* delivery_time_days: Time taken for delivery.
* customer_rating: Customer satisfaction score on a scale of 1–5.
* returned: A boolean indicator (True/False) of whether the order was returned.
* order_date: The date the transaction occurred.


# **Part 2 Preprocessing Pipeline:**
1. Load raw dataset
2. Inspect data structure
3. Identify missing values, duplicates, inconsistent labels, invalid values, and outliers
4. Standardize column names and categorical values
5. Convert data types and order_date
6. Handle invalid values and outliers
7. Impute missing values
8. Create new variables
9. Validate final cleaned dataset
10. Export cleaned CSV

Import libraries and load data

In [ ]:

import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)
pd.set_option("display.width", 1200)

file_path = "ecommerce_customer_behavior_raw.csv"

df_raw = pd.read_csv(file_path)
df = df_raw.copy()

print("Dataset imported successfully.")
print("Number of rows and columns:", df.shape)

Dataset imported successfully.
Number of rows and columns: (5000, 10)



# **A. RAW DATA INSPECTION**



## **I. Inspect dataset structure**
Inspect column names, sample rows, data types, and overall dataset structure.

In [ ]:
print("Column names:")
print(df.columns.tolist())

print("\nFirst 5 rows:")
display(df.head())

print("\nLast 5 rows:")
display(df.tail())

print("\nData types:")
display(df.dtypes)

print("\nDataset information:")
df.info()

Column names:
['order_id', 'customer_age', 'customer_gender', 'product_category', 'payment_method', 'order_value_usd', 'delivery_time_days', 'customer_rating', 'returned', 'order_date']

First 5 rows:


,order_id,customer_age,customer_gender,product_category,payment_method,order_value_usd,delivery_time_days,customer_rating,returned,order_date
0,1,37,Other,Home & Living,credit card,1366.60,10.0,2.3,No,01-01-22
1,2,41,Female,Fashion,credit card,1175.89,6.0,1.8,Yes,02-01-22
2,3,30,Other,Sports,credit card,1283.97,7.0,3.2,No,03-01-22
3,4,58,Female,Electronics,credit card,1245.76,14.0,3.8,No,04-01-22
4,5,59,Other,Fashion,credit card,1411.06,7.0,3.2,No,05-01-22



Last 5 rows:


,order_id,customer_age,customer_gender,product_category,payment_method,order_value_usd,delivery_time_days,customer_rating,returned,order_date
4995,4996,18,Female,Groceries,Credit Card,420.44,7.0,1.3,Yes,05-09-35
4996,4997,19,Male,Sports,UPI,1389.97,2.0,1.2,Yes,06-09-35
4997,4998,42,Male,Sports,Credit Card,845.66,4.0,1.3,Yes,07-09-35
4998,4999,46,Other,Sports,Credit Card,1077.51,3.0,2.6,Yes,08-09-35
4999,5000,60,Male,Sports,UPI,290.47,8.0,1.5,Yes,09-09-35



Data types:


,0
order_id,int64
customer_age,int64
customer_gender,object
product_category,object
payment_method,object
order_value_usd,float64
delivery_time_days,float64
customer_rating,float64
returned,object
order_date,object



Dataset information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   order_id            5000 non-null   int64  
 1   customer_age        5000 non-null   int64  
 2   customer_gender     5000 non-null   object 
 3   product_category    5000 non-null   object 
 4   payment_method      4950 non-null   object 
 5   order_value_usd     5000 non-null   float64
 6   delivery_time_days  4926 non-null   float64
 7   customer_rating     4900 non-null   float64
 8   returned            5000 non-null   object 
 9   order_date          5000 non-null   object 
dtypes: float64(3), int64(2), object(5)
memory usage: 390.8+ KB


**RESULTS:**
The dataset has 5,000 rows and 10 columns. Some variables such as payment_method, returned, and order_date are stored as object type.


## **II. Summary statistics**
Generate descriptive statistics for numerical and categorical variables.

In [ ]:
print("Summary statistics for numerical variables:")
display(df.describe())

print("\nSummary statistics for categorical variables:")
display(df.describe(include="object"))

Summary statistics for numerical variables:


,order_id,customer_age,order_value_usd,delivery_time_days,customer_rating
count,5000.000000,5000.000000,5000.000000,4926.000000,4900.000000
mean,2501.537400,43.257000,757.882900,7.505887,3.021224
std,1442.926134,15.013662,519.912692,4.164665,1.159620
min,1.000000,18.000000,5.570000,1.000000,1.000000
25%,1251.750000,30.000000,367.087500,4.000000,2.000000
50%,2500.500000,43.000000,756.820000,7.000000,3.000000
75%,3751.250000,56.000000,1119.487500,11.000000,4.000000
max,5000.000000,69.000000,10000.000000,40.000000,8.000000



Summary statistics for categorical variables:


,customer_gender,product_category,payment_method,returned,order_date
count,5000,5000,4950,5000,5000
unique,3,8,8,4,4966
top,Male,Beauty,UPI,Yes,15-08-30
freq,1711,854,1014,2511,6


**Some abnormal values are detected, such as order_value_usd = 10000, delivery_time_days = 40, and customer_rating = 8.**

## **III. Check missing values**
Check the number and percentage of missing values in each column

In [ ]:
missing_summary = pd.DataFrame({
    "missing_values": df.isnull().sum(),
    "missing_percentage": (df.isnull().sum() / len(df) * 100).round(2)
})

print("Missing value summary:")
display(missing_summary)

Missing value summary:


,missing_values,missing_percentage
order_id,0,0.00
customer_age,0,0.00
customer_gender,0,0.00
product_category,0,0.00
payment_method,50,1.00
order_value_usd,0,0.00
delivery_time_days,74,1.48
customer_rating,100,2.00
returned,0,0.00
order_date,0,0.00


RESULT

---


| Column               | Missing values |
| -------------------- | -------------: |
| `payment_method`     |             50 |
| `delivery_time_days` |             74 |
| `customer_rating`    |            100 |


## **IV. Check duplicate records**

In [ ]:
duplicate_rows = df.duplicated().sum()

print("Number of duplicate rows:", duplicate_rows)

if "order_id" in df.columns:
    duplicate_order_id = df["order_id"].duplicated().sum()
    print("Number of duplicated order_id:", duplicate_order_id)

Number of duplicate rows: 29
Number of duplicated order_id: 29


## **V. Inspect categorical variables**
Inspect categorical variables to detect inconsistent labels and capitalization issues.

In [ ]:
categorical_columns_raw = df.select_dtypes(include=["object"]).columns

for col in categorical_columns_raw:
    print("\n" + "-" * 60)
    print("Column:", col)
    print("Number of unique values:", df[col].nunique(dropna=False))
    print(df[col].value_counts(dropna=False))


------------------------------------------------------------
Column: customer_gender
Number of unique values: 3
customer_gender
Male      1711
Other     1648
Female    1641
Name: count, dtype: int64

------------------------------------------------------------
Column: product_category
Number of unique values: 8
product_category
Beauty           854
Fashion          847
Sports           847
Groceries        832
Home & Living    803
Electronics      797
electronics       10
ELECTRONICS       10
Name: count, dtype: int64

------------------------------------------------------------
Column: payment_method
Number of unique values: 9
payment_method
UPI                 1014
Wallet               995
Credit Card          993
Debit Card           979
Cash on Delivery     938
NaN                   50
credit card           11
CREDIT CARD           10
Credit card           10
Name: count, dtype: int64

------------------------------------------------------------
Column: returned
Number of unique v

Inconsistent labels are detected, such as Electronics, electronics, ELECTRONICS; Yes, yes, No, NO.

## **VI. Inspect numerical ranges**
Check minimum, maximum, mean, and missing values of numerical variables.

In [ ]:
numerical_columns_raw = df.select_dtypes(include=["int64", "float64"]).columns

for col in numerical_columns_raw:
    print("\n" + "-" * 60)
    print("Column:", col)
    print("Minimum:", df[col].min())
    print("Maximum:", df[col].max())
    print("Mean:", round(df[col].mean(), 2))
    print("Missing:", df[col].isnull().sum())


------------------------------------------------------------
Column: order_id
Minimum: 1
Maximum: 5000
Mean: 2501.54
Missing: 0

------------------------------------------------------------
Column: customer_age
Minimum: 18
Maximum: 69
Mean: 43.26
Missing: 0

------------------------------------------------------------
Column: order_value_usd
Minimum: 5.57
Maximum: 10000.0
Mean: 757.88
Missing: 0

------------------------------------------------------------
Column: delivery_time_days
Minimum: 1.0
Maximum: 40.0
Mean: 7.51
Missing: 74

------------------------------------------------------------
Column: customer_rating
Minimum: 1.0
Maximum: 8.0
Mean: 3.02
Missing: 100


**Some values require treatment: order_value_usd has a maximum of 10000, delivery_time_days has a maximum of 40, and customer_rating has a maximum of 8.**

## **VII. Check invalid values and outliers**
Check invalid values and detect outliers using the IQR method.

In [ ]:
print("Invalid value checks:")

invalid_age = df[(df["customer_age"] < 18) | (df["customer_age"] > 100)]
invalid_order_value = df[df["order_value_usd"] <= 0]
invalid_delivery_time = df[
    (df["delivery_time_days"] <= 0) | (df["delivery_time_days"] > 30)
]
invalid_rating = df[
    (df["customer_rating"] < 1) | (df["customer_rating"] > 5)
]

print("Invalid customer_age count:", len(invalid_age))
print("Invalid order_value_usd count:", len(invalid_order_value))
print("Invalid delivery_time_days count:", len(invalid_delivery_time))
print("Invalid customer_rating count:", len(invalid_rating))


print("\nPotential outlier detection using IQR method:")

numeric_cols_for_outlier = [
    "customer_age",
    "order_value_usd",
    "delivery_time_days",
    "customer_rating"
]

outlier_summary = []

for col in numeric_cols_for_outlier:
    series = df[col].dropna()

    Q1 = series.quantile(0.25)
    Q3 = series.quantile(0.75)
    IQR = Q3 - Q1

    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR

    number_of_outliers = df[
        (df[col] < lower_bound) | (df[col] > upper_bound)
    ].shape[0]

    outlier_summary.append({
        "variable": col,
        "Q1": Q1,
        "Q3": Q3,
        "IQR": IQR,
        "lower_bound": lower_bound,
        "upper_bound": upper_bound,
        "number_of_outliers": number_of_outliers
    })

outlier_summary_df = pd.DataFrame(outlier_summary)
display(outlier_summary_df)

Invalid value checks:
Invalid customer_age count: 0
Invalid order_value_usd count: 0
Invalid delivery_time_days count: 5
Invalid customer_rating count: 5

Potential outlier detection using IQR method:


,variable,Q1,Q3,IQR,lower_bound,upper_bound,number_of_outliers
0,customer_age,30.0000,56.0000,26.0,-9.0000,95.0000,0
1,order_value_usd,367.0875,1119.4875,752.4,-761.5125,2248.0875,5
2,delivery_time_days,4.0000,11.0000,7.0,-6.5000,21.5000,5
3,customer_rating,2.0000,4.0000,2.0,-1.0000,7.0000,5


RESULT

---
| Issue                            | Count |
| -------------------------------- | ----: |
| Invalid `delivery_time_days`     |     5 |
| Invalid `customer_rating`        |     5 |
| Outliers in `order_value_usd`    |     5 |
| Outliers in `delivery_time_days` |     5 |
| Outliers in `customer_rating`    |     5 |



## **VIII. Inspect date format**
Inspect the format of the order_date column before converting it to datetime

In [ ]:
print("Order_date sample values before conversion:")
display(df["order_date"].head(20))

print("\nData type of order_date before conversion:")
print(df["order_date"].dtype)

Order_date sample values before conversion:


,order_date
0,01-01-22
1,02-01-22
2,03-01-22
3,04-01-22
4,05-01-22
5,06-01-22
6,07-01-22
7,08-01-22
8,09-01-22
9,10-01-22



Data type of order_date before conversion:
object


**The order_date column contains mixed formats such as 1/1/2022 and 13-01-22, so a custom conversion function is needed**

## **IX. Raw data quality summary**
Summarize raw data quality issues and their impact on business analysis.

RAW DATA QUALITY ISSUE SUMMARY

Main data quality issues:
1. Missing values are found in payment_method, delivery_time_days, and customer_rating.
2. Duplicate records and duplicated order_id values are present.
3. Categorical variables contain inconsistent labels due to different capitalization.
4. order_date is stored as object/text and has mixed date formats.
5. Some numerical variables contain invalid or extreme values:
   - customer_rating values outside the valid 1-5 range
   - unrealistic delivery_time_days values
   - extremely high order_value_usd values

Business impact:
- Missing values can reduce the accuracy of customer and delivery analysis.
- Duplicate records can overstate transaction counts, revenue, and return rates.
- Inconsistent category labels can cause incorrect group comparison.
- Invalid values and outliers can distort averages and business insights.
- Incorrect date formats can affect trend analysis and time-based visualization.

# **B. DATA PREPROCESSING**

## **I. Standardize column names**

In [ ]:
df.columns = (
    df.columns
    .str.strip()
    .str.lower()
    .str.replace(" ", "_")
)

print("Column names after standardization:")
print(df.columns.tolist())

Column names after standardization:
['order_id', 'customer_age', 'customer_gender', 'product_category', 'payment_method', 'order_value_usd', 'delivery_time_days', 'customer_rating', 'returned', 'order_date']


**Column names are standardized to lowercase with underscores**

## **II. Remove duplicate records**
Remove duplicate records so each transaction appears only once.

In [ ]:
rows_before_duplicates = len(df)

df = df.drop_duplicates()

rows_after_duplicates = len(df)

print("Rows before removing duplicates:", rows_before_duplicates)
print("Rows after removing duplicates:", rows_after_duplicates)
print("Number of duplicate rows removed:", rows_before_duplicates - rows_after_duplicates)

print("Duplicated order_id after removing duplicate rows:", df["order_id"].duplicated().sum())

Rows before removing duplicates: 5000
Rows after removing duplicates: 4971
Number of duplicate rows removed: 29
Duplicated order_id after removing duplicate rows: 0


## **III. Standardize categorical variables**
Standardize categorical variables to remove inconsistent capitalization.

In [ ]:
def clean_text(value):
    if pd.isna(value):
        return np.nan
    return str(value).strip().title()

categorical_columns = [
    "customer_gender",
    "product_category",
    "payment_method",
    "returned"
]

for col in categorical_columns:
    df[col] = df[col].apply(clean_text)

df["payment_method"] = df["payment_method"].replace({
    "Upi": "UPI",
    "Cash On Delivery": "Cash on Delivery"
})

print("Unique values after standardizing categorical variables:")

for col in categorical_columns:
    print("\n" + col)
    print(df[col].value_counts(dropna=False))

Unique values after standardizing categorical variables:

customer_gender
customer_gender
Male      1699
Other     1643
Female    1629
Name: count, dtype: int64

product_category
product_category
Fashion          845
Beauty           845
Sports           843
Groceries        829
Electronics      811
Home & Living    798
Name: count, dtype: int64

payment_method
payment_method
Credit Card         1020
UPI                 1008
Wallet               989
Debit Card           972
Cash on Delivery     932
NaN                   50
Name: count, dtype: int64

returned
returned
Yes    2510
No     2461
Name: count, dtype: int64


**Values such as electronics, ELECTRONICS are standardized to Electronics; yes, NO are standardized to Yes, No.**

## **IV. Convert order_date to datetime**
Convert order_date from text/object to datetime for time-based analysis.

In [ ]:
def parse_mixed_order_date(value):
    if pd.isna(value):
        return pd.NaT

    value = str(value).strip()

    if "-" in value:
        return pd.to_datetime(value, errors="coerce", dayfirst=True)

    if "/" in value:
        return pd.to_datetime(value, errors="coerce", dayfirst=False)

    return pd.to_datetime(value, errors="coerce")

df["order_date"] = df["order_date"].apply(parse_mixed_order_date)

print("order_date after conversion:")
print("Data type:", df["order_date"].dtype)
print("Minimum date:", df["order_date"].min())
print("Maximum date:", df["order_date"].max())
print("Missing dates after conversion:", df["order_date"].isnull().sum())

print("\nSample converted dates:")
display(df[["order_id", "order_date"]].head(20))

order_date after conversion:
Data type: datetime64[ns]
Minimum date: 2022-01-01 00:00:00
Maximum date: 2035-09-09 00:00:00
Missing dates after conversion: 0

Sample converted dates:


,order_id,order_date
0,1,2022-01-01
1,2,2022-01-02
2,3,2022-01-03
3,4,2022-01-04
4,5,2022-01-05
5,6,2022-01-06
6,7,2022-01-07
7,8,2022-01-08
8,9,2022-01-09
9,10,2022-01-10


RESULT

---

Data type: datetime64[ns]
Minimum date: 2022-01-01
Maximum date: 2035-12-08
Missing dates after conversion: 0

## **V. Handle invalid numerical values**
Convert invalid numerical values into missing values before imputation.

In [ ]:
invalid_rating_mask = (df["customer_rating"] < 1) | (df["customer_rating"] > 5)
print("Invalid customer_rating values converted to NaN:", invalid_rating_mask.sum())
df.loc[invalid_rating_mask, "customer_rating"] = np.nan

invalid_delivery_mask = (df["delivery_time_days"] <= 0) | (df["delivery_time_days"] > 30)
print("Invalid delivery_time_days values converted to NaN:", invalid_delivery_mask.sum())
df.loc[invalid_delivery_mask, "delivery_time_days"] = np.nan

invalid_value_mask = df["order_value_usd"] <= 0
print("Invalid order_value_usd values converted to NaN:", invalid_value_mask.sum())
df.loc[invalid_value_mask, "order_value_usd"] = np.nan

Invalid customer_rating values converted to NaN: 5
Invalid delivery_time_days values converted to NaN: 5
Invalid order_value_usd values converted to NaN: 0


## **VI. Cap outliers in order_value_usd**
Reduce the influence of extreme order_value_usd values using IQR capping

In [ ]:
Q1 = df["order_value_usd"].quantile(0.25)
Q3 = df["order_value_usd"].quantile(0.75)
IQR = Q3 - Q1

upper_bound = Q3 + 1.5 * IQR

outlier_order_value_count = (df["order_value_usd"] > upper_bound).sum()

print("Order value upper bound using IQR:", round(upper_bound, 2))
print("Order value outliers capped:", outlier_order_value_count)

df["order_value_usd"] = np.where(
    df["order_value_usd"] > upper_bound,
    upper_bound,
    df["order_value_usd"]
)

Order value upper bound using IQR: 2251.81
Order value outliers capped: 5


## **VII. Handle missing values**
Impute missing values using suitable methods: mode for categorical variables and median for numerical variables.

In [ ]:
print("Missing values before imputation:")
display(df.isnull().sum())

payment_mode = df["payment_method"].mode(dropna=True)[0]
df["payment_method"] = df["payment_method"].fillna(payment_mode)
print("Missing payment_method filled with mode:", payment_mode)

delivery_median = df["delivery_time_days"].median()
df["delivery_time_days"] = df["delivery_time_days"].fillna(delivery_median)
print("Missing delivery_time_days filled with median:", delivery_median)

rating_median = df["customer_rating"].median()
df["customer_rating"] = df["customer_rating"].fillna(rating_median)
print("Missing customer_rating filled with median:", rating_median)

missing_dates = df["order_date"].isnull().sum()
print("Missing order_date after date conversion:", missing_dates)

if missing_dates > 0:
    df = df.dropna(subset=["order_date"])

print("\nMissing values after imputation:")
display(df.isnull().sum())

Missing values before imputation:


,0
order_id,0
customer_age,0
customer_gender,0
product_category,0
payment_method,50
order_value_usd,0
delivery_time_days,79
customer_rating,105
returned,0
order_date,0


Missing payment_method filled with mode: Credit Card
Missing delivery_time_days filled with median: 7.0
Missing customer_rating filled with median: 3.0
Missing order_date after date conversion: 0

Missing values after imputation:


,0
order_id,0
customer_age,0
customer_gender,0
product_category,0
payment_method,0
order_value_usd,0
delivery_time_days,0
customer_rating,0
returned,0
order_date,0


**RESULT**

---

| Column               | Method               |
| -------------------- | -------------------- |
| `payment_method`     | Mode = `Credit Card` |
| `delivery_time_days` | Median = `7.0`       |
| `customer_rating`    | Median = `3.0`       |


## **VIII. Convert data types after cleaning**
Convert variables into appropriate data types after cleaning.

In [ ]:
df["delivery_time_days"] = df["delivery_time_days"].round().astype(int)
df["customer_age"] = df["customer_age"].astype(int)
df["order_id"] = df["order_id"].astype(int)

print("Data types after cleaning:")
display(df.dtypes)

Data types after cleaning:


,0
order_id,int64
customer_age,int64
customer_gender,object
product_category,object
payment_method,object
order_value_usd,float64
delivery_time_days,int64
customer_rating,float64
returned,object
order_date,datetime64[ns]


# **C. FEATURE ENGINEERING**



## **I. Create returned_flag**
Convert returned from Yes/No to 1/0 for return-rate analysis

In [ ]:
df["returned_flag"] = df["returned"].map({
    "Yes": 1,
    "No": 0
})

print("Returned flag distribution:")
display(df["returned_flag"].value_counts(dropna=False))

print("\nOverall return rate:")
print(round(df["returned_flag"].mean() * 100, 2), "%")

Returned flag distribution:


,count
returned_flag,
1,2510
0,2461



Overall return rate:
50.49 %


## **II. Create age_group**
Segment customers into age groups for customer behavior analysis.

In [ ]:
age_bins = [17, 25, 35, 45, 55, 70]
age_labels = ["18-25", "26-35", "36-45", "46-55", "56-69"]

df["age_group"] = pd.cut(
    df["customer_age"],
    bins=age_bins,
    labels=age_labels
)

print("Age group distribution:")
display(df["age_group"].value_counts().sort_index())

Age group distribution:


,count
age_group,
18-25,782
26-35,964
36-45,992
46-55,918
56-69,1315


## **III. Create order_value_segment**
Classify orders into low, medium, and high value segments.

In [ ]:
df["order_value_segment"] = pd.qcut(
    df["order_value_usd"],
    q=3,
    labels=["Low value", "Medium value", "High value"]
)

print("Order value segment distribution:")
display(df["order_value_segment"].value_counts())

Order value segment distribution:


,count
order_value_segment,
Low value,1657
Medium value,1657
High value,1657


**This variable supports return-rate comparison by spending level.**

## **IV. Create delivery_speed**
Classify delivery time into fast, standard, and slow delivery.

In [ ]:
def categorize_delivery_speed(days):
    if days <= 3:
        return "Fast delivery"
    elif days <= 7:
        return "Standard delivery"
    else:
        return "Slow delivery"

df["delivery_speed"] = df["delivery_time_days"].apply(categorize_delivery_speed)

print("Delivery speed distribution:")
display(df["delivery_speed"].value_counts())

Delivery speed distribution:


,count
delivery_speed,
Slow delivery,2423
Standard delivery,1495
Fast delivery,1053


**The delivery_speed variable supports analysis of delivery performance, rating, and return rate**

## **V. Create time-based variables**
Create time-based variables for yearly, monthly, and quarterly trend analysis

In [ ]:
df["order_year"] = df["order_date"].dt.year
df["order_month"] = df["order_date"].dt.month
df["order_quarter"] = df["order_date"].dt.quarter
df["order_month_year"] = df["order_date"].dt.to_period("M").astype(str)

print("Time-based variables created:")
display(df[[
    "order_date",
    "order_year",
    "order_month",
    "order_quarter",
    "order_month_year"
]].head())

Time-based variables created:


,order_date,order_year,order_month,order_quarter,order_month_year
0,2022-01-01,2022,1,1,2022-01
1,2022-01-02,2022,1,1,2022-01
2,2022-01-03,2022,1,1,2022-01
3,2022-01-04,2022,1,1,2022-01
4,2022-01-05,2022,1,1,2022-01


**RESULT**

---
| Variable           | Purpose                              |
| ------------------ | ------------------------------------ |
| `order_year`       | Yearly trend analysis                |
| `order_month`      | Monthly trend analysis               |
| `order_quarter`    | Quarterly trend analysis             |
| `order_month_year` | Month-year time-series visualization |



## **VI. Create is_high_value_order**
Identify high-value orders based on the 75th percentile of order_value_usd

In [ ]:
high_value_threshold = df["order_value_usd"].quantile(0.75)

df["is_high_value_order"] = np.where(
    df["order_value_usd"] >= high_value_threshold,
    1,
    0
)

print("High-value order threshold:", round(high_value_threshold, 2))

print("\nHigh-value order distribution:")
display(df["is_high_value_order"].value_counts())

High-value order threshold: 1120.86

High-value order distribution:


,count
is_high_value_order,
0,3728
1,1243


## **VII. New variable summary**

In [ ]:
new_variables = pd.DataFrame({
    "New variable": [
        "returned_flag",
        "age_group",
        "order_value_segment",
        "delivery_speed",
        "order_year",
        "order_month",
        "order_quarter",
        "order_month_year",
        "is_high_value_order"
    ],
    "Purpose": [
        "Convert return status from Yes/No to 1/0 for return-rate analysis",
        "Group customers by age for customer segmentation",
        "Classify orders into low, medium, and high value groups",
        "Classify delivery time into fast, standard, and slow delivery",
        "Extract year from order_date for yearly trend analysis",
        "Extract month from order_date for monthly trend analysis",
        "Extract quarter from order_date for quarterly trend analysis",
        "Create month-year period for time-series visualization",
        "Identify high-value orders based on the 75th percentile"
    ]
})

display(new_variables)

,New variable,Purpose
0,returned_flag,Convert return status from Yes/No to 1/0 for r...
1,age_group,Group customers by age for customer segmentation
2,order_value_segment,"Classify orders into low, medium, and high val..."
3,delivery_speed,"Classify delivery time into fast, standard, an..."
4,order_year,Extract year from order_date for yearly trend ...
5,order_month,Extract month from order_date for monthly tren...
6,order_quarter,Extract quarter from order_date for quarterly ...
7,order_month_year,Create month-year period for time-series visua...
8,is_high_value_order,Identify high-value orders based on the 75th p...


**9 new variables are created to support EDA and managerial analysis**

# **D. FINAL VALIDATION**




## **I. Final dataset validation**
Check shape, data types, missing values, and duplicates after preprocessing.

In [ ]:
print("Final dataset shape:")
print(df.shape)

print("\nFinal data types:")
display(df.dtypes)

print("\nMissing values after preprocessing:")
display(df.isnull().sum())

print("\nDuplicate rows after preprocessing:")
print(df.duplicated().sum())

print("Duplicated order_id after preprocessing:", df["order_id"].duplicated().sum())

print("\nFinal cleaned dataset preview:")
display(df.head())

Final dataset shape:
(4971, 19)

Final data types:


,0
order_id,int64
customer_age,int64
customer_gender,object
product_category,object
payment_method,object
order_value_usd,float64
delivery_time_days,int64
customer_rating,float64
returned,object
order_date,datetime64[ns]



Missing values after preprocessing:


,0
order_id,0
customer_age,0
customer_gender,0
product_category,0
payment_method,0
order_value_usd,0
delivery_time_days,0
customer_rating,0
returned,0
order_date,0



Duplicate rows after preprocessing:
0
Duplicated order_id after preprocessing: 0

Final cleaned dataset preview:


,order_id,customer_age,customer_gender,product_category,payment_method,order_value_usd,delivery_time_days,customer_rating,returned,order_date,returned_flag,age_group,order_value_segment,delivery_speed,order_year,order_month,order_quarter,order_month_year,is_high_value_order
0,1,37,Other,Home & Living,Credit Card,1366.60,10,2.3,No,2022-01-01,0,36-45,High value,Slow delivery,2022,1,1,2022-01,1
1,2,41,Female,Fashion,Credit Card,1175.89,6,1.8,Yes,2022-01-02,1,36-45,High value,Standard delivery,2022,1,1,2022-01,1
2,3,30,Other,Sports,Credit Card,1283.97,7,3.2,No,2022-01-03,0,26-35,High value,Standard delivery,2022,1,1,2022-01,1
3,4,58,Female,Electronics,Credit Card,1245.76,14,3.8,No,2022-01-04,0,56-69,High value,Slow delivery,2022,1,1,2022-01,1
4,5,59,Other,Fashion,Credit Card,1411.06,7,3.2,No,2022-01-05,0,56-69,High value,Standard delivery,2022,1,1,2022-01,1


**RESULT**

---

Final dataset shape: (4971, 19)
Missing values: 0
Duplicate rows: 0
Duplicated order_id: 0

## **II. Final invalid value checks**
Ensure that no invalid values remain after cleaning.

In [ ]:
print("Final invalid value checks:")

print("Invalid customer_age count:",
      len(df[(df["customer_age"] < 18) | (df["customer_age"] > 100)]))

print("Invalid order_value_usd count:",
      len(df[df["order_value_usd"] <= 0]))

print("Invalid delivery_time_days count:",
      len(df[(df["delivery_time_days"] <= 0) | (df["delivery_time_days"] > 30)]))

print("Invalid customer_rating count:",
      len(df[(df["customer_rating"] < 1) | (df["customer_rating"] > 5)]))

Final invalid value checks:
Invalid customer_age count: 0
Invalid order_value_usd count: 0
Invalid delivery_time_days count: 0
Invalid customer_rating count: 0


## **III. Simulated dataset note**

In [ ]:
max_year = df["order_date"].dt.year.max()

print("Dataset date range note:")

if max_year > 2026:
    print("""
The order_date column extends beyond the current year.
This suggests that the dataset is likely a simulated academic dataset.
Therefore, the analysis should focus on preprocessing quality, customer behavior patterns,
return behavior, and managerial interpretation rather than real-time business performance.
""")
else:
    print("""
The order_date range does not extend far beyond the current year.
The dataset can be used for trend analysis within the observed period.
""")

Dataset date range note:

The order_date column extends beyond the current year.
This suggests that the dataset is likely a simulated academic dataset.
Therefore, the analysis should focus on preprocessing quality, customer behavior patterns,
return behavior, and managerial interpretation rather than real-time business performance.



**The dataset is treated as a simulated academic dataset.**

**EDA READINESS CHECK**





**Return rate by product category**\
Check whether the data is ready for return-rate analysis by product category.

In [ ]:
return_by_category = (
    df.groupby("product_category")["returned_flag"]
    .mean()
    .sort_values(ascending=False)
    .mul(100)
    .round(2)
)

print("Return rate by product category:")
display(return_by_category)

Return rate by product category:


,returned_flag
product_category,
Beauty,52.19
Groceries,51.51
Home & Living,51.25
Sports,50.53
Electronics,49.32
Fashion,48.17


**Return rates can be compared across product categories.**

**Return rate by payment method**\
Check whether the data is ready for return-rate analysis by payment method.

In [ ]:
return_by_payment = (
    df.groupby("payment_method")["returned_flag"]
    .mean()
    .sort_values(ascending=False)
    .mul(100)
    .round(2)
)

print("Return rate by payment method:")
display(return_by_payment)

Return rate by payment method:


,returned_flag
payment_method,
UPI,52.98
Cash on Delivery,51.61
Credit Card,50.00
Debit Card,49.28
Wallet,48.63


**Return rates can be compared across payment methods.**

**Rating by delivery speed**\
Check whether the data is ready for rating analysis by delivery speed.

In [ ]:
rating_by_delivery = (
    df.groupby("delivery_speed")["customer_rating"]
    .mean()
    .round(2)
)

print("Average customer rating by delivery speed:")
display(rating_by_delivery)

Average customer rating by delivery speed:


,customer_rating
delivery_speed,
Fast delivery,3.09
Slow delivery,2.99
Standard delivery,3.01


**The relationship between delivery speed and customer rating can be analyzed.**

**Return rate by customer and order segments**\
Check whether the data is ready for return-rate analysis by age group and order value segment.

In [ ]:
return_by_age_group = (
    df.groupby("age_group", observed=False)["returned_flag"]
    .mean()
    .mul(100)
    .round(2)
)

print("Return rate by age group:")
display(return_by_age_group)


return_by_order_value_segment = (
    df.groupby("order_value_segment", observed=False)["returned_flag"]
    .mean()
    .mul(100)
    .round(2)
)

print("\nReturn rate by order value segment:")
display(return_by_order_value_segment)

Return rate by age group:


,returned_flag
age_group,
18-25,50.13
26-35,50.93
36-45,52.02
46-55,49.67
56-69,49.81



Return rate by order value segment:


,returned_flag
order_value_segment,
Low value,51.18
Medium value,50.15
High value,50.15


**The dataset is ready for return behavior analysis by customer characteristics and order value.**

**Final preprocessing summary**

FINAL PREPROCESSING OUTCOME SUMMARY

After preprocessing, the dataset contains 4971 rows and 19 columns.

Final outcomes:
1. Column names are standardized.
2. Duplicate records were removed.
3. Missing values were handled.
4. Inconsistent categorical labels were standardized.
5. order_date was converted into datetime format.
6. Invalid ratings and unrealistic delivery times were handled.
7. Extreme order values were capped using the IQR method.
8. returned was transformed into returned_flag for quantitative analysis.
9. New variables were created to support customer behavior analysis.
10. The cleaned dataset is ready for EDA, visualization, and managerial analysis.

The final cleaned dataset is stored in the variable named df.
The original raw dataset is stored in the variable named df_raw.

**The final dataset contains 4,971 rows and 19 columns, with no missing values, duplicates, or invalid values.**

**EXPORT CLEANED DATASET**




**Export cleaned CSV**\
Export the cleaned dataset as a CSV file for submission or further analysis.

In [ ]:
output_file = "ecommerce_customer_behavior_cleaned.csv"

df.to_csv(output_file, index=False)

print("Cleaned dataset has been exported successfully.")
print("File name:", output_file)

Cleaned dataset has been exported successfully.
File name: ecommerce_customer_behavior_cleaned.csv


**Download cleaned CSV from Colab**

In [ ]:
from google.colab import files

files.download(output_file)

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>